xgboost_ray/main.py

```diff
@@ -810,7 +826,37 @@ def train(params: Dict,
             "`dtrain = RayDMatrix(data=data, label=label)`.".format(
                 type(dtrain)))
 
-    _try_add_tune_callback(kwargs)
+    if not ray.is_initialized():
+        ray.init()
+
+    gpus_per_actor = ray_params.gpus_per_actor
+    cpus_per_actor = ray_params.cpus_per_actor
+
+    # Automatically set gpus_per_actor if left at the default value
+    if gpus_per_actor == -1:
+        gpus_per_actor = 0
+        if "tree_method" in params and params["tree_method"].startswith("gpu"):
+            gpus_per_actor = 1
+
+    # Automatically set cpus_per_actor if left at the default value
+    # Will be set to the number of cluster CPUs divided by the number of
+    # actors, bounded by the minimum number of CPUs across actors nodes.
+    if cpus_per_actor <= 0:
+        cluster_cpus = _ray_get_cluster_cpus() or 1
+        cpus_per_actor = min(
+            int(_get_min_node_cpus() or 1),
```

Comment:
Can we move all the code of 831 to 852 to a separate function like:

```
cpus_per_actor, gpus_per_actor = _validate_resources(...)
```

Reply:
I moved into a separate function